src/testing/systest/supervisor/nemesis.zig

```diff
@@ -0,0 +1,290 @@
+
+pub fn init(
+    allocator: std.mem.Allocator,
+    random: std.rand.Random,
+    replicas: []*LoggedProcess,
+) !*Self {
+    assert(replicas.len <= replicas_count_max);
+
+    const nemesis = try allocator.create(Self);
+    errdefer allocator.destroy(nemesis);
+
+    nemesis.* = .{
+        .allocator = allocator,
+        .random = random,
+        .replicas = replicas,
+        .netem_rules = netem.Rules{ .delay = null, .loss = null },
+    };
+
+    return nemesis;
+}
+
+pub fn deinit(self: *Self) void {
+    self.network_netem_delete_all() catch {};
+    const allocator = self.allocator;
+    allocator.destroy(self);
+}
+
+const Havoc = enum {
+    terminate_replica,
+    restart_replica,
+    network_delay_add,
+    network_delay_remove,
+    network_loss_add,
+    network_loss_remove,
+    sleep,
+};
+
+pub fn wreak_havoc(self: *Self) !bool {
+    const havoc = weighted(self.random, Havoc, .{
+        .sleep = 50,
+        .terminate_replica = 1,
+        .restart_replica = 10,
+        .network_delay_add = 1,
+        .network_delay_remove = 10,
+        .network_loss_add = 1,
+        .network_loss_remove = 10,
+    });
+    switch (havoc) {
+        .terminate_replica => return try self.terminate_replica(),
+        .restart_replica => return try self.restart_replica(),
+        .network_delay_add => {
+            if (self.netem_rules.delay != null) {
+                return false;
+            }
+            const delay_ms = self.random.intRangeAtMost(u16, 10, 200);
+            self.netem_rules.delay = .{
+                .time_ms = delay_ms,
+                .jitter_ms = @divFloor(delay_ms, 10),
+                .correlation_pct = 75,
+            };
+            try self.netem_sync();
+            return true;
+        },
+        .network_delay_remove => {
+            if (self.netem_rules.delay == null) {
+                return false;
+            }
+            self.netem_rules.delay = null;
+            try self.netem_sync();
+            return true;
+        },
+        .network_loss_add => {
+            if (self.netem_rules.loss != null) {
+                return false;
+            }
+            const loss_pct = self.random.intRangeAtMost(u8, 5, 100);
+            self.netem_rules.loss = .{
+                .loss_pct = loss_pct,
+                .correlation_pct = 75,
+            };
+            try self.netem_sync();
+            return true;
+        },
+        .network_loss_remove => {
+            if (self.netem_rules.loss == null) {
+                return false;
+            }
+            self.netem_rules.loss = null;
+            try self.netem_sync();
+            return true;
+        },
+        .sleep => {
+            std.time.sleep(3 * std.time.ns_per_s);
+            return true;
+        },
+    }
+}
+
+fn random_replica_in_state(self: *Self, state: LoggedProcess.State) ?*LoggedProcess {
+    var matching: [replicas_count_max]*LoggedProcess = undefined;
+    var pos: u8 = 0;
```

Comment:
1. We don't generally contract words, so, `position` (curiously, I used to be a `let mut res` person before joining TB, but I am in the `var result` camp now!)
2. We don't usually use `position`, it is rather `index` or `count`